router/handler_edge_ctrl/sessionAdded.go

```diff
@@ -17,42 +17,249 @@
 package handler_edge_ctrl
 
 import (
+	"errors"
 	"github.com/golang/protobuf/proto"
 	"github.com/michaelquigley/pfxlog"
 	"github.com/openziti/edge/controller/env"
-	"github.com/openziti/edge/router/internal/fabric"
+	"github.com/openziti/edge/controller/sync_strats"
 	"github.com/openziti/edge/pb/edge_ctrl_pb"
+	"github.com/openziti/edge/router/internal/fabric"
 	"github.com/openziti/foundation/channel2"
+	"sync"
+	"time"
 )
 
 type sessionAddedHandler struct {
-	sm fabric.StateManager
+	sm          fabric.StateManager
+	syncTracker *sessionSyncTracker
+	reqChan     chan *sessionAddedWithState
+	syncReady   chan []*edge_ctrl_pb.Session
+	syncFail    chan error
 }
 
 func NewSessionAddedHandler(sm fabric.StateManager) *sessionAddedHandler {
-	return &sessionAddedHandler{
-		sm: sm,
+	handler := &sessionAddedHandler{
+		sm:        sm,
+		reqChan:   make(chan *sessionAddedWithState, 100),
+		syncReady: make(chan []*edge_ctrl_pb.Session, 0),
+		syncFail:  make(chan error, 0),
 	}
+
+	go handler.startRecieveSync()
+	go handler.startSyncApplier()
+	go handler.startSyncFail()
+
+	return handler
 }
 
 func (h *sessionAddedHandler) ContentType() int32 {
 	return env.SessionAddedType
 }
 
-func (h *sessionAddedHandler) HandleReceive(msg *channel2.Message, _ channel2.Channel) {
+func (h *sessionAddedHandler) HandleReceive(msg *channel2.Message, ch channel2.Channel) {
 	go func() {
 		req := &edge_ctrl_pb.SessionAdded{}
 		if err := proto.Unmarshal(msg.Body, req); err == nil {
-			for _, session := range req.Sessions {
-				pfxlog.Logger().Debugf("received new session %+v", session)
-				h.sm.AddSession(session)
-			}
-
 			if req.IsFullState {
-				h.sm.RemoveMissingSessions(req.Sessions)
+
+				reqWithState := &sessionAddedWithState{
+					SessionAdded: req,
+				}
+
+				if syncStrategyType, syncState, err := parseInstantSyncHeaders(msg); err == nil {
+					reqWithState.SyncStrategyType = syncStrategyType
+					reqWithState.InstantSyncState = syncState
+				} else {
+					pfxlog.Logger().WithField("strategy", syncStrategyType).WithField("msgContentType", msg.ContentType).WithError(err).Errorf("sync headers not present (old controller) or only partial present(error), treading as legacy: %v", err)
+				}
+
+				h.reqChan <- reqWithState
+			} else {
+				for _, session := range req.Sessions {
+					h.sm.AddSession(session)
+				}
 			}
 		} else {
-			pfxlog.Logger().Panic("could not convert message as network session added")
+			pfxlog.Logger().Panic("could not convert message as session added")
 		}
 	}()
 }
+
+func (h *sessionAddedHandler) startSyncApplier() {
+	for sessions := range h.syncReady {
+		for _, session := range sessions {
+			h.sm.AddSession(session)
+		}
+		h.sm.RemoveMissingSessions(sessions)
+
+		pfxlog.Logger().Infof("finished sychronizing sessions [count: %d]", len(sessions))
+	}
+}
+
+func (h *sessionAddedHandler) startSyncFail() {
```

Comment:
Wasn't sure if we needed this based on @plorenz 's work.

If we do need resync logic for sessions, we do it here.

Reply:
I added a reverse sync. The router will check that the sessions it is tracking are still valid.